src/ch/ch_monitor.c

```diff
@@ -563,25 +564,338 @@ chMonitorBuildSocketCmd(virDomainObjPtr vm, const char *socket_path)
     return cmd;
 }
 
+static const char *virCHMonitorEventStrings[] = {
+    "vmm:starting",
+    "vmm:shutdown",
+    "vm:booting", "vm:booted",
+    "vm:pausing", "vm:paused",
+    "vm:resuming", "vm:resumed",
+    "vm:snapshotting", "vm:snapshotted",
+    "vm:restoring", "vm:restored",
+    "vm:resizing", "vm:resized",
+    "vm:shutdown", "vm:deleted",
+    "cpu_manager:create_vcpu",
+    "virtio-device:activated", "virtio-device:reset"
+};
+
+static virCHMonitorEvent virCHMonitorEventFromString(const char *sourceStr,
+        const char *eventStr)
+{
+    int i;
+    g_auto(virBuffer) buf = VIR_BUFFER_INITIALIZER;
+    const char *event;
+
+    virBufferAsprintf(&buf, "%s:%s", sourceStr, eventStr);
+    event = virBufferCurrentContent(&buf);
+
+    for (i = 0; i < virCHMonitorEventMax; i++) {
+        if STREQ(event, virCHMonitorEventStrings[i])
+            break;
+    }
+
+    return i;
+}
+
+static int virCHMonitorValidateEventsJSON(char *buffer, size_t sz,
+            bool *incomplete)
+{
+    /*
+     * Marks the start of a JSON doc(starting with '{')
+     */
+    char *json_start = buffer;
+    /*
+     * Marks the start of the buffer where the scan starts.
+     * It could be either:
+     *      - Start of the buffer. Or
+     *      - Next location after a valid JSON doc.
+     */
+    char *scan_start = buffer;
+    int blocks = 0;
+    int events = 0;
+    int i = 0;
+
+    if (sz == 0)
+        return 0;
+
+    /*
+     * Check if the message is a wellformed JSON. Try to find all
+     * wellformed JSON doc and adjust the buffer accordingly by
+     * removing invalid snippets in the buffer.
+     */
+    do {
+        if (buffer[i] == '{') {
+            blocks++;
+
+            if (blocks != 1)
+                continue;
+
+            /*
+             * Possible start of a valid JSON doc. Check if
+             * there were any white characters or garbage
+             * before the JSON doc at this location.
+             */
+            json_start = buffer + i;
+            if (scan_start != json_start) {
+                int invalid_chars = json_start - scan_start;
+                VIR_WARN("invalid json or white chars in buffer: %.*s",
+                         invalid_chars, scan_start);
+                memmove(scan_start, json_start, invalid_chars);
```

Comment:
I'm not so sure about this `memmove`. It copies `invalid_chars` number of items from the valid part of the string over the invalid part which doesn't seem useful. What about the following:
```
memmove(buffer, json_start, sz - invalid_chars);
memset(buffer + sz - invalid_chars, 0, invalid_chars);
json_start = 0;
scan_start = 0;
i = 0;
sz -= invalid_chars;
```
This copies the contents from the start of the json string to the end of the buffer (skipping over the invalid characters essentially) to the start of the buffer, then zeroes out the end of the buffer for however many invalid characters there were (and then resets the various tracking offsets).

What do you think about this approach?

Reply:
Sorry, what I wanted was actually move rest of the buffer from json_start to scan_start. The exact idea that you proposed above was my intention, but I messed it up ;-)

Moving to start of the buffer might not always be correct. The invalid chars can be between two valid json blocks as well. So we need to move only those areas. A slightly modified version of your proposal:
```
memmove(scan_start, json_start, sz - i);
memset(scan_start + sz - i, 0, sz - invalid_chars);
json_start = scan_start;
i -= invalid_chars;
sz -= invalid_chars;
```
How about this?

Reply:
Ahh you are certainly correct, I wasn't thinking about the invalid characters between full blocks. Your proposal looks right to me.